### PR TITLE
Content-Keys Type Definitions & key->id derivation functions

### PR DESCRIPTION
**Pull request by victor-wei126**
This PR does a few things:

- Implements the type definitions for the state and history network `content-keys`. The associated methods for SSZ encoding and decoding are defined as well
- Implements the key->id derivation functions for both state and history networks. Currently, these are associated functions of the corresponding type def. There is another wrapper function which determines if a key is for state/history and chooses the appropriate derivation function to call
- Separate `types.rs` in `trin-core` into separate files since it is getting large.

After this PR is merged, I plan to:
- Move the u256 and u512 from mod.rs to `types/utils.rs`
- Implement cross-client test vectors

### CODE REVIEW

**Comment by victor-wei126** [comment]
@mrferris I've gotten a derivation function for state. I used SSZ encode & decode traits and basically it matches the decoded key to perform the appropriate function. Let me know if this is where we're trying to head

File: trin-state/src/utils.rs
Comment (carver) [inline trin-state/src/utils.rs:30]: Looks like the derive method should return a `Result` instead. Crashing the program on an invalid content key sounds bad.

File: trin-state/src/utils.rs
Comment (carver) [inline trin-state/src/utils.rs:35]: Along the same lines: shouldn't crash on a bad utf-8.

File: trin-state/src/utils.rs
Comment (carver) [inline trin-state/src/utils.rs:83]: Ah, interesting, I may have missed this convo. Are we really switching to sha3_256 proper? There's a lot of room for confusion, for it to live alongside keccak256. Wondering if the spec was just a little loose in saying sha, when it meant keccak.

File: trin-history/src/utils.rs
Comment (carver) [inline trin-history/src/utils.rs:41]: nit: it's nice to end the file with a newline

File: trin-history/src/utils.rs
Comment (carver) [inline trin-history/src/utils.rs:36]: Feels silly to redefine this in every crate. Seems like this and keccak could go into the `trin-core` utils.

File: trin-history/src/utils.rs
Comment (carver) [inline trin-history/src/utils.rs:17]: It's not immediately obvious to me that using `+` on a U256 will concatenate the bytes instead of numerically add the results. In fact, I would have guessed the opposite, but I guess that's just more of a complaint with the U256 lib. If there's another way to be more explicit that we are concatenating bytes, just for readability in trin, I think that would be preferable.

File: trin-core/src/portalnet/types.rs
Comment (carver) [inline trin-core/src/portalnet/types.rs:354]: I worry that this duplicated name can be a bit confusing. Maybe something like this?
```suggestion
  AccountTrieNodeKey(AccountTrieNode),
```

File: trin-core/src/portalnet/types.rs
Comment (carver) [inline trin-core/src/portalnet/types.rs:368]: Have you looked at the errors resulting from this? It seems like it might be useful to explicitly mention that the code was attempting to decode with `AccountTrieNode`, maybe maybe that's already clear enough in `e`? Also, if `e` doesn't include the byte source, I'd like to see it added here, hex-encoded. It will be so much easier to track down bugs later from the logs, if we have the source bytes and know which thing we were trying to decode as.

File: trin-core/src/portalnet/types.rs
Comment (carver) [inline trin-core/src/portalnet/types.rs:389]: This message be the only text you see in a spew of logs, when trying to debug a problem. Without more context, "No content type" wouldn't help me find the problem easily. Maybe something like this would help a bit more?
```suggestion
          Err("Cannot decode StateContentKey from empty bytes".to_string())
```

File: trin-core/src/portalnet/types.rs
Comment (carver) [inline trin-core/src/portalnet/types.rs:386]: Similarly, would recommend including the `content_type` byte that has no match in this message (and include that you're looking in the `StateContentKey`).

**Review by carver — COMMENTED** [review]
I think the code would really benefit from some tests of decoding content keys, and converting to IDs, for both successful cases and error cases.

Feel free to ping me for another review when tests are added (and existing tests are passing).

File: trin-state/src/utils.rs
Comment (victor-wei126) [reply trin-state/src/utils.rs:83]: The crate I am using has both sha3_256 and Keccak256. It notes that Keccak256 is the NIST submission without padding changes. So I'm guessing sha3_256 is the actual "controversial" standard? The specs for state and history specify both "sha256" and just "keccak" so my guess is they mean the SHA-2 family for "sha256" and "keccak" as the "unproper". @pipermerriam

File: trin-history/src/utils.rs
Comment (victor-wei126) [reply trin-history/src/utils.rs:17]: I was under the impression that the `+` was adding the U256s, which is what I thought we wanted? I'm not sure since the spec is not clear on this point. I also add two U256s in calculating the ContractStorageTrieProof key, specified by the spec [here](https://github.com/ethereum/portal-network-specs/blob/master/state-network.md#contract-storage-trie-proof). The only thing I have to support that `+` is actually addition and not concatenation is that in the state-network spec `|` is used for concatenation, but `+` symbol is used, so I assumed its addition. History spec is unclear.

File: trin-core/src/portalnet/types.rs
Comment (victor-wei126) [reply trin-core/src/portalnet/types.rs:354]: idk why I added key to everything else except for state keys, probably cuz i thought the names were getting too long lol. will add

File: trin-history/src/utils.rs
Comment (carver) [reply trin-history/src/utils.rs:17]: Oh, interesting! I just hadn't noticed that it was actually addition in some places. Will U256 automatically roll-over as the spec defines? More things to test. I'm also not clear why we are sometimes using addition and sometimes concatenation. There are fewer question marks with concat (like not having to worry about rollover), so I would recommend pushing back on the spec a little, if there's no obviously good reason to use addition.

**Comment by carver** [comment]
Can you fix the formatting errors so we can see the tests passing?

File: trin-history/src/utils.rs
Comment (mrferris) [inline trin-history/src/utils.rs:11]: I'm thinking that these should be methods of the `ContentKey` enums defined in `types.rs` instead of external `util` functions, and then I'll update`PortalStorage` methods take a `ContentKey` and just call `derive_content_id` on it internally instead of having to pass in the function like we're doing now. That will be simpler and better organized.

Any reason you can see that we shouldn't do that?

File: trin-core/src/utils/content_key.rs
Comment (mrferris) [inline trin-core/src/utils/content_key.rs:15]: nit: needs a newline at end of file

File: src/main.rs
Comment (mrferris) [inline src/main.rs:15]: not for this PR, but I don't love `utils::utils`. Once you merge this I'm going to make a `utils/db.rs` that will include this function, and then we can think of a better name for a file containing the other things.

**Review by mrferris — COMMENTED** [review]
Looks pretty good, I left a few thoughts, some of which are related to each other.

Looks like you need a `cargo fmt` too.

File: trin-core/src/portalnet/types.rs
Comment (victor-wei126) [reply trin-core/src/portalnet/types.rs]: I think you mean `FixedVector`? Since this needs 20 bytes and 20 bytes only. Also, since we can now start using `ssz_types` this raises the larger question of if we want to replace all types with `ssz_types` implemented in the crate? For example, for a bytes32, we can use a bare rust type of `[u8; 32]` or use `ssz_types` of `FixedVector<u8, typenum::U32`. Having the bare type makes things simpler but using the ssz_type gives us additional flexibility.

File: trin-history/src/utils.rs
Comment (victor-wei126) [reply trin-history/src/utils.rs:11]: This means you can get rid of the closure function in PortalStorage right? Yeah 100% I think this has multiple benefits. The only thing is, since it is now a method, we have to call it like `key.derive_content_id()`, where `key` is of type `ContentKey`, as I understand. So we'd need to get the type first, which means instead of passing a byte array to `derive_content_id` and then decoding in that method, we would have to decode the byte array first somewhere outside of the method, which gives us the type, and then we can call the derivation function. But regardless, I think this way of doing things is much cleaner and organized.

Comment (mrferris) [reply trin-history/src/utils.rs:11]: 👍 I'd think you can just do a `match` on the first byte inside the function but I might not be seeing what you mean fully.

Either way, let's go that route. Once you merge this in I'll refactor PortalStorage to take the `ContentKey` type you define in this PR.

File: trin-core/src/cli.rs
Comment (pipermerriam) [inline trin-core/src/cli.rs:1]: A good practice is to keep strong separation between organizational changes and functionality changes.  A good way to deal with this is to extract just the part that moved `HexData` to this new location and then open a pull request with just that change.  Then you can rebase this branch on top of that other PR which will allow a reviewer of this PR to not have to sort through more changed lines than absolutely necessary.

**Comment by victor-wei126** [comment]
> I would prefer to see the content key derivation fns split out into trin-state/ and trin-history/ in this pr.

@njgheorghita Just to confirm I'm understanding you correctly, do you mean move just the derivation functions out of trin-core, or also move the entire `StateContentKey` to `trin-state/` and same for history? If anything, I think if we were to move things out of trin-core, then we should move the entire enum and all it's methods, not just the derivation function.

But I would like to push back a bit and say that by moving these type definitions out of trin-core and into their respective crates, we then have to make trin-state and trin-history a dependency of trin-core (unless there is some way around this that I'm unaware of), since `PortalStorage` relies on this functionality, and that is in trin-core. So my opinion as of now is to just keep it everything in trin-core, and if the specific subnetworks need the functionality we can move it out.

File: trin-core/src/portalnet/mod.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/mod.rs:20]: So, this was the home for `U256` - but it doesn't feel like the appropriate place for these uint definitions/tests. Personally, I don't think we should have any code inside our `mod.rs` files - it should be strictly for `pub mod ..` declarations. How do you feel about moving `U256` and `U512` definitions and tests over to `trin-core/src/types/uint.rs` or something similar?

File: trin-core/src/portalnet/mod.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/mod.rs:128]: *nitpick* this and the following test name should begin with `test_*`

File: trin-core/src/portalnet/mod.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/mod.rs:135]: Sorry for going full boomer, and tbf I do this kinda thing constantly when I'm developing locally, but I'd rather not push any meme code to master if you can replace this test with something more boring.

File: trin-core/src/portalnet/mod.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/mod.rs:149]: It doesn't seem to me like this is the correct expected case (and for the test below)? 
- eg. shouldn't the expected case (even though it'll always panic) be `U256([U64::MAX;4])`?

It's not obvious to me what's throwing the panic here - is it line 153 or 154? If it's 154, it would be better to not panic and `assert_eq` explicitly against the `U512toU256Error::Overflow` that's being returned for overflows.

File: trin-core/src/portalnet/types.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/types.rs:564]: I'd rather see this move happen in this pr. I can't think of a great reason to delay the move for a later pr (but i'm all ears). But yes, in regards to your other comment, it makes sense to keep them inside `trin-core` if `trin-core` depends on these types.

File: trin-core/src/portalnet/types.rs
Comment (njgheorghita) [inline trin-core/src/portalnet/types.rs:1055]: I don't think this is ideal for our tests. It seems possible that these might not run at all if the `let` statement doesn't match. At the very least you should add an `else { panic!(..) }` statement to the end. Or you could do something like [here](https://stackoverflow.com/questions/34953711/unwrap-inner-type-when-enum-variant-is-known) where you implement functions that'll return the inner value

**Review by njgheorghita — COMMENTED** [review]
Great! It feels like this is almost there - just a couple nitpicks / re-org suggestions. It makes sense to me to handle the re-org in this pr and not save it for a later pr. 

I'd like to go through this one more time though after the next round of fixes, just to go through the test cases in a little more detail - though everything checks out at first glance. If you could also squash / rebase the majority of those commits, we don't want those ending up in master.

Also, it seems like some cross-client test vectors for the encoding / decoding would be extremely useful (like https://github.com/ethereum/portal-network-specs/blob/dbbb4c8774/portal-wire-test-vectors.md). Is that something you could pick up after this pr? The other test vectors were great for pointing out some blind spots in our code.

File: trin-core/src/portalnet/mod.rs
Comment (victor-wei126) [reply trin-core/src/portalnet/mod.rs:20]: Yeah that makes sense to me since those are types. Should I address that in this PR?

File: trin-core/src/portalnet/mod.rs
Comment (victor-wei126) [reply trin-core/src/portalnet/mod.rs:135]: haha yes i will change that

File: trin-core/src/portalnet/mod.rs
Comment (njgheorghita) [reply trin-core/src/portalnet/mod.rs:20]: Probably not this pr since there's a lot going on already, but i'd just take a not to address it in a follow-up pr after this gets merged

File: trin-core/src/portalnet/mod.rs
Comment (victor-wei126) [reply trin-core/src/portalnet/mod.rs:149]: Well in my impl if there is an overflow the conversion just fails and doesn't default to the max u256 value, so I changed this test case to have an expected value of `Ut12toU256Error::Overflow` as that should be the actual expected value. Let me know if this is what you were thinking.

File: trin-core/src/portalnet/types.rs
Comment (victor-wei126) [reply trin-core/src/portalnet/types.rs]: A general testing question: we're expecting this test to succeed, since I made sure to have the proper input, but is including a panic just for safekeeping, in case the input was wrong somehow?

Comment (victor-wei126) [reply trin-core/src/portalnet/types.rs]: Also, I just reviewed the code again and I think adding an `else { panic!(..) }` at the end of the `if` would be redundant because the `unwrap()` above will cause the test to panic in the case of decoding failure, which is what the else would be doing. The `if let` statement handles the case of successful decoding.

Comment (njgheorghita) [reply trin-core/src/portalnet/types.rs]: Yeah including the panic is just for safekeeping. In this example, let's say the deserializing code changes somehow that instead of returning `AccountTrieNodeKey` for `decoded` - it returns `ContractStorageTrieNodeKey`. The `unwrap()` wouldn't catch this, b/c it was successfully deserialized. Then the if statement never gets run and the test passes successfully. Generally, I think it's smart to avoid `if` statements or other logic in your tests unless it's strictly necessary. In this scenario, adding a little `else {panic!()}` clause guarantees that either the tested assumptions inside the `if()` statement are run - or it panics.

**Review by njgheorghita — APPROVED** [review]
Looks great! Thanks for keeping with this and handling all the suggestions. Sorry for the late response here, it fell through the cracks a bit on my end. In the future, if you're waiting on a review from me for more than a couple days - feel free to re-ping me here or on discord.

Just one comment to address re: testing & panics. Also curious if you've put any time towards creating test vectors for these messages (like [these](https://github.com/ethereum/portal-network-specs/blob/master/portal-wire-test-vectors.md)? If not, creating a small issue to track that would be neat.

**Pull request merged**
